Threaded third-party snapshots and reports

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR modifies `extras/create_snapshots_reports_scorecard.py` so that third-party snapshots and report are generated in threads, in a similar manner to how "regular" (non-third-party) snapshots and reports are generated.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

As the number of third-party reports has grown, the time it takes to create these reports has also grown:
```console
2024-03-24 17:29:05,238 INFO - Time to create all grouping node and third-party snapshots: 41.35 minutes
2024-03-24 18:24:06,500 INFO - Time to create all third-party reports: 55.02 minutes
2024-03-24 18:26:49,450 INFO -   Third-party snapshots generated: 32
2024-03-24 18:26:49,451 INFO -   Third-party reports generated: 32
```

By implementing multi-threading, we should be able to significantly reduce the amount of time it takes to generate third-party snapshots and reports each week.

This also gets us one small step closer to containerizing all of our snapshot and report generation.

Resolves #60.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##
 
To verify that these changes worked as expected, I first commented out any parts of `create_snapshots_reports_scorecard.py` that were not necessary for my local testing or related to the changes that I made (e.g. pausing/resuming the commander, creating the sample report, etc).  Then I replaced the [`snapshot_command`](https://github.com/cisagov/cyhy-reports/blob/cb0adf6744ab93520e5f737e357dac9866803666/extras/create_snapshots_reports_scorecard.py#L289) and [`report_command`](https://github.com/cisagov/cyhy-reports/blob/cb0adf6744ab93520e5f737e357dac9866803666/extras/create_snapshots_reports_scorecard.py#L575-L583) with harmless test and error commands like this:
```python: 
    # Force a failure for testing purposes 1% of the time.  This is much higher
    # than our actual failure rate (which is essentially 0% these days), but it
    # allows us to test the failure handling code.
    random_error = random.randint(1, 100)
    if random_error == 50:
        snapshot_command = ["sh", "foo"]
    else:
        random_sleep = random.randint(1, 3)
        snapshot_command = [
            "sleep",
            str(random_sleep),
        ]
```

Once I had the script set up so that it could be run without actually generating any snapshots or reports (each thread only running my test/error commands above), I executed a full run using all of the entities in the production database with read-only credentials, just in case (e.g. `./create_snapshots_reports_scorecard.py --no-dock --no-log --no-pause cyhy-read-only scan-read-only`).  I confirmed that the output of the full run looked as expected (output below sanitized and shortened):

```console
2024-03-28 15:21:57,412 INFO - Number of snapshots generated: 7139
2024-03-28 15:21:57,413 INFO -   Number of third-party and grouping node snapshots generated: 32
2024-03-28 15:21:57,413 INFO - Number of snapshots failed: 62
2024-03-28 15:21:57,413 INFO -   Number of third-party and grouping node snapshots failed: 0
2024-03-28 15:21:57,414 ERROR - Failed snapshots (reports not attempted):
2024-03-28 15:21:57,414 ERROR - AAAAA
2024-03-28 15:21:57,415 ERROR - BBBBB
<snip>
2024-03-28 15:21:57,430 ERROR - ZZZZZ
2024-03-28 15:21:57,431 INFO - Number of reports generated: 7973
2024-03-28 15:21:57,431 INFO -   Third-party reports generated: 32
2024-03-28 15:21:57,431 INFO - Number of reports failed: 100
2024-03-28 15:21:57,431 INFO -   Third-party reports failed: 0
2024-03-28 15:21:57,431 ERROR - Failed reports:
2024-03-28 15:21:57,431 ERROR - AAAA
2024-03-28 15:21:57,431 ERROR - BBBB
<snip>
2024-03-28 15:21:57,441 ERROR - ZZZZ
2024-03-28 15:21:57,441 INFO - Snapshot performance:
2024-03-28 15:21:57,446 INFO -   Longest snapshot: 3.2 seconds (0.1 minutes)
2024-03-28 15:21:57,450 INFO -   Median snapshot: 2.0 seconds (0.0 minutes)
2024-03-28 15:21:57,450 INFO -   Mean snapshot: 2.1 seconds (0.0 minutes)
2024-03-28 15:21:57,451 INFO -   Shortest snapshot: 0.0 seconds (0.0 minutes)
2024-03-28 15:21:57,455 INFO - Longest snapshots:
2024-03-28 15:21:57,456 INFO -   AAA: 3.2 seconds (0.1 minutes)
2024-03-28 15:21:57,456 INFO -   BBB: 3.2 seconds (0.1 minutes)
2024-03-28 15:21:57,456 INFO -   CCC: 3.2 seconds (0.1 minutes)
2024-03-28 15:21:57,456 INFO -   DDD: 3.2 seconds (0.1 minutes)
2024-03-28 15:21:57,456 INFO -   EEE: 3.2 seconds (0.1 minutes)
2024-03-28 15:21:57,456 INFO -   FFF: 3.2 seconds (0.1 minutes)
2024-03-28 15:21:57,456 INFO -   GGG: 3.2 seconds (0.1 minutes)
2024-03-28 15:21:57,456 INFO -   HHH: 3.2 seconds (0.1 minutes)
2024-03-28 15:21:57,456 INFO -   III: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,456 INFO -   JJJ: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,456 INFO - Third-party and grouping node snapshot performance:
2024-03-28 15:21:57,456 INFO -   Longest third-party/grouping node snapshot: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,456 INFO -   Median third-party/grouping node snapshot: 2.1 seconds (0.0 minutes)
2024-03-28 15:21:57,456 INFO -   Mean third-party/grouping node snapshot: 2.1 seconds (0.0 minutes)
2024-03-28 15:21:57,457 INFO -   Shortest third-party/grouping node snapshot: 1.0 seconds (0.0 minutes)
2024-03-28 15:21:57,457 INFO - Longest third-party/grouping node snapshots:
2024-03-28 15:21:57,457 INFO -   LLL: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,457 INFO -   MMM-RR: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,457 INFO -   NNN: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,457 INFO -   OOO: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,457 INFO -   PPP: 3.0 seconds (0.1 minutes)
2024-03-28 15:21:57,457 INFO -   QQQ: 3.0 seconds (0.1 minutes)
2024-03-28 15:21:57,457 INFO -   RRR: 3.0 seconds (0.1 minutes)
2024-03-28 15:21:57,457 INFO -   SSS: 3.0 seconds (0.1 minutes)
2024-03-28 15:21:57,457 INFO -   TTT: 3.0 seconds (0.1 minutes)
2024-03-28 15:21:57,457 INFO -   UUU-RR: 2.1 seconds (0.0 minutes)
2024-03-28 15:21:57,457 INFO - Report performance:
2024-03-28 15:21:57,458 INFO -   Longest report: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,459 INFO -   Median report: 2.0 seconds (0.0 minutes)
2024-03-28 15:21:57,460 INFO -   Mean report: 2.1 seconds (0.0 minutes)
2024-03-28 15:21:57,461 INFO -   Shortest report: 0.0 seconds (0.0 minutes)
2024-03-28 15:21:57,466 INFO - Longest reports:
2024-03-28 15:21:57,466 INFO -   AAA: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,466 INFO -   BBB: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,466 INFO -   CCC: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,466 INFO -   DDD: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,466 INFO -   EEE: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,466 INFO -   FFF: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,466 INFO -   GGG: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,466 INFO -   HHH: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,466 INFO -   III: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,466 INFO -   JJJ: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,466 INFO - Third-party report performance:
2024-03-28 15:21:57,466 INFO -   Longest third-party report: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,466 INFO -   Median third-party report: 2.0 seconds (0.0 minutes)
2024-03-28 15:21:57,466 INFO -   Mean third-party report: 1.9 seconds (0.0 minutes)
2024-03-28 15:21:57,466 INFO -   Shortest third-party report: 1.0 seconds (0.0 minutes)
2024-03-28 15:21:57,466 INFO - Longest third-party reports:
2024-03-28 15:21:57,467 INFO -   LLL: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,467 INFO -   MMM: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,467 INFO -   NNN: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,467 INFO -   OOO: 3.1 seconds (0.1 minutes)
2024-03-28 15:21:57,467 INFO -   PPP: 3.0 seconds (0.1 minutes)
2024-03-28 15:21:57,467 INFO -   QQQ: 3.0 seconds (0.1 minutes)
2024-03-28 15:21:57,467 INFO -   RRR: 3.0 seconds (0.1 minutes)
2024-03-28 15:21:57,467 INFO -   SSS: 3.0 seconds (0.1 minutes)
2024-03-28 15:21:57,467 INFO -   TTT: 2.1 seconds (0.0 minutes)
2024-03-28 15:21:57,467 INFO -   UUU: 2.1 seconds (0.0 minutes)
2024-03-28 15:21:57,467 INFO - Time to generate snapshots: 27.94 minutes
2024-03-28 15:21:57,467 INFO - Time to generate reports: 8.91 minutes
2024-03-28 15:21:57,467 INFO - Time to generate grouping node snapshots: 0.92 minutes
2024-03-28 15:21:57,467 INFO - Time to generate third-party snapshots: 0.98 minutes
2024-03-28 15:21:57,467 INFO - Time to generate third-party reports: 0.27 minutes
2024-03-28 15:21:57,467 INFO - Total time: 38.10 minutes
2024-03-28 15:21:57,467 INFO - END
```

I also reviewed the log output to ensure that when snapshots and reports failed (for both regular and third-party entities), the code correctly handled the failures.  For example, when a regular snapshot fails, any descendant snapshots that should've been created by that parent snapshot were also marked as failed.  However, that should not happen for third-party snapshots since third-party snapshots do not create descendant snapshots- they only use previously-created snapshots.

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Deploy this change to Production
- [x] Closely review the output of this script the first time it runs in Production to ensure everything worked as expected

dav3r committed:
Add global list variables and thread locks for third-party snapshots and reports

dav3r committed:
Simplify descendants query

This works since we just grabbed the correct list of report-receiving orgs in an earlier query in create_list_of_reports_to_generate().

dav3r committed:
Make create_list_of_reports_to_generate() third-party-report-aware

dav3r committed:
Make generate_snapshot() third-party-snapshot-aware

dav3r committed:
Make manage_snapshot_threads() third-party-snapshot-aware

dav3r committed:
Make generate_snapshots_from_list() third-party-snapshot-aware

dav3r committed:
Add new function prepare_for_third_party_snapshots()

dav3r committed:
Make generate_report() third-party-report-aware

dav3r committed:
Make generate_reports_from_list() third-party-report-aware

dav3r committed:
Make manage_report_threads() third-party-report-aware

dav3r committed:
Rename a variable and add another

dav3r committed:
Update calls to functions that generate regular (non-third-party) snapshots and reports

Also, improve comments and logging a bit.

dav3r committed:
Create third-party snapshots and reports using third-party-aware functions

dav3r committed:
Clean up and improve summary logging at end of script

This includes adding log output for performance details of third-party snapshot and report generation.

dav3r committed:
Delete code and variables that are now obsolete

jsf9k reviewed (approved):
This was a big lift!  It all looks correct to me.

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -217,17 +238,26 @@ def sample_report(cyhy_db_section, scan_db_section, nolog):
+        # generating a report since it would be empty.
+        query = {
+            "report_types": REPORT_TYPE.CYHY_THIRD_PARTY,
+            "children": {"$exists": True, "$ne": []},
```
Alphabetization police.

```suggestion
            "children": {"$exists": True, "$ne": []},
            "report_types": REPORT_TYPE.CYHY_THIRD_PARTY,
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -337,46 +387,138 @@ def generate_snapshots_from_list(db, cyhy_db_section):
+                "_id": {"$in": list(all_tp_descendants)},
+                "stakeholder": False,
+                "report_types": [],
+                "scan_types": [],
```
```suggestion
                "report_types": [],
                "scan_types": [],
                "stakeholder": False,
```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -337,46 +387,138 @@ def generate_snapshots_from_list(db, cyhy_db_section):
+    # snapshots_to_generate because we are still single-threaded at this point
+
+    if third_party:
+        # Some extra preparation must be done before generating third-party snapshots
```
Might be nice to have some logging here to indicate which path we went down since the `else` logs.

dav3r replied:
I didn't log anything there because it starts logging inside of `prepare_for_third_party_snapshots()`, but I can add something here just in case.

mcdonnnj replied:
mm yeah I neglected to account for logging in `prepare_for_third_party_snapshots()`. That should be fine.

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -924,60 +955,50 @@ def main():
         sample_report(
             cyhy_db_section, scan_db_section, nolog
         )  # Create the sample (anonymized) report
         
```
This is a bunch of whitespace I noticed. Hoping the suggestions doesn't remove the actual line.

```suggestion

```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -1046,27 +1071,82 @@ def main():
             logging.info("Longest snapshots:")
             for i in snapshot_durations[:10]:
                 logging.info("  %s: %.1f seconds (%.1f minutes)", i[0], i[1], i[1] / 60)
+            
```
More unnecessary whitespace.

```suggestion

```

mcdonnnj commented on extras/create_snapshots_reports_scorecard.py:
```diff
@@ -79,6 +79,27 @@
+# locks) are reused for third-party snapshots and reports, so we don't need
+# variables and locks for them below.
+successful_tp_snapshots = list()
+stps_lock = threading.Lock()
```
Since third-party work is done after first-party work we could probably reuse a number of these locks and just do

```python
with lock_in_question:
    if third_party:
        # Do third-party thing
    else:
        # Do first-party thing
```

Just a thought since adding a single boolean value check to the critical section is negligible in my mind.

dav3r replied:
I see your point.  I created those third-party locks mainly to reduce confusion and try to keep things named consistently, but since I'm already re-using some of the things (`snapshots_to_generate` and `reports_to_generate`), what's a few more?  I will ponder.

dav3r replied:
I have come to the conclusion that I agree with your suggestion, which I put in https://github.com/cisagov/cyhy-reports/pull/118/commits/dea1135e1c791a79494dd906799d262b439fa564.  Excellent thinking!

mcdonnnj reviewed (approved):
I think this looks ok. Some minor suggestions for your consideration.

dav3r committed:
Alphabetize fields in query

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Remove extraneous whitespace

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add back unintentionally removed newlines

These newlines were unexpectedly removed (along with whitespace that was meant to be removed) in https://github.com/cisagov/cyhy-reports/pull/118/commits/bffccb3d3f6cf146e3f9ca7160bf31f81da3712b

GitHub's "commit suggestion" feature is a bit overzealous when it comes to whitespace removal.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Get rid of third-party locks

There's no benefit to having separate locks for third-party snapshots and reports.  It makes more sense to just use the same locks for all snapshots and reports since third-party work is not concurrent with non-third-party work.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r commented:
I confirmed that this new code worked as expected during the most-recent weekly reporting run.  There were three third-party reports that failed because they require very large amounts of memory, which was not available while the reports were being generated simultaneously.  If the changes documented in #107 are implemented such that the re-tries are single-threaded (no concurrency), then that should alleviate this problem.  I plan to implement such a change this week.
